### PR TITLE
[doc][hackathon] To add Adagrad Optimizer to the documentation

### PR DESCRIPTION
**Pull request by iramazanli**
It has been discussed before that adding description of Optimization algorithms to PyTorch Core documentation may result in a nice Optimization research tutorial. In the following tracking issue we mentioned about all the necessary algorithms and links to the originally published paper  https://github.com/pytorch/pytorch/issues/63236.

In this PR we are adding description of Adagrad to the documentation.  For more details, we refer to the paper 
http://jmlr.org/papers/v12/duchi11a.html

AdaGradAlgo

cc @vincentqb @iramazanli

### CODE REVIEW

**Comment by albanD** [comment]
I don't see initial_accumulator_value and lr_decay?

Also should we mention what is the behavior when the gradient is sparse?

**Comment by iramazanli** [comment]
> I don't see initial_accumulator_value and lr_decay?
> 
> Also should we mention what is the behavior when the gradient is sparse?

* The reason initial_accumulator_value is skipped is because it is skipped in Args section here :  https://github.com/pytorch/pytorch/blob/f79df24859125fca8e8af799fb2be44dc3293752/torch/optim/adagrad.py#L12 which i tried to keep consistent with pseudocode.

* The reason I didn't add lr_decay is little bit conceptual.  Indeed any optimization algorithm could be called with any kind of learning rate schedulers, and for each type of scheduler look of pseudocode documented here might change. In particular, i choose \gamma instead of \gamma_t  in all these algorithms to illustrate focus on optimization algorithm rather other learning rate / scheduler.

Let me know if you have feedbacks regarding these two point.

Regarding sparse behavior, i would be happy to follow your suggestion.

**Comment by albanD** [comment]
> The reason initial_accumulator_value is skipped is because it is skipped in Args section here :

Is that done on purpose? It looks like the args doc need to be updated.

In any case, in the rendered version, the argument is present in the signature, so I think it should be documented properly.

> The reason I didn't add lr_decay is little bit conceptual.

I agree with your point that we might want to discourage people from using this argument.
But I feel like the goal of these pseudo code is to explain what the implementation is doing. And so it should reflect all the arguments.
If we want to ask people not to use this (or even deprecate it) that can be mentioned in the args doc below.

**Comment by iramazanli** [comment]
> > The reason initial_accumulator_value is skipped is because it is skipped in Args section here :
> 
> Is that done on purpose? It looks like the args doc need to be updated.
> 
> In any case, in the rendered version, the argument is present in the signature, so I think it should be documented properly.
> 
> > The reason I didn't add lr_decay is little bit conceptual.
> 
> I agree with your point that we might want to discourage people from using this argument.
> But I feel like the goal of these pseudo code is to explain what the implementation is doing. And so it should reflect all the arguments.
> If we want to ask people not to use this (or even deprecate it) that can be mentioned in the args doc below.

It's done ! thanks for the comments.

File: torch/optim/adagrad.py [hunk 1]
```diff
@@ -4,9 +4,30 @@
...
+       \begin{aligned}
+            &\rule{110mm}{0.4pt}                                                                 \\
+            &\textbf{input}      : \gamma \text{ (lr)}, \: \theta_0 \text{ (params)}, \: f(\theta)
+                \text{ (objective)}, \: \lambda \text{ (weight decay)},                          \\
+            &\hspace{12mm}    \tau \text{ (initial accumulator value)}, \: \eta\text{ (lr decay)}\\
+            &\textbf{initialize} :  G_0 \leftarrow \tau                                   \\[-1.ex]
+            &\rule{110mm}{0.4pt}                                                                 \\
+            &\textbf{for} \: t=1 \: \textbf{to} \: \ldots \: \textbf{do}                         \\
+            &\hspace{5mm}g_t           \leftarrow   \nabla_{\theta} f_t (\theta_{t-1})           \\
+            &\hspace{5mm}\gamma    \leftarrow \gamma / (1 +(t-1) \eta)                           \\
```
Comment (albanD) [inline torch/optim/adagrad.py:19]: nit you don't want to assign to gamma here right? The new value should not be used by the next step right?

File: torch/optim/adagrad.py [hunk 2]
```diff
@@ -4,9 +4,30 @@
...
+            &\hspace{5mm} \textbf{if} \: \lambda \neq 0                                          \\
+            &\hspace{10mm} g_t \leftarrow g_t + \lambda \theta_{t-1}                             \\
+            &\hspace{5mm}G_t       \leftarrow  G_{t-1} + g_t g_t^T                               \\
```
Comment (albanD) [inline torch/optim/adagrad.py:22]: Isn't this g_t^2 ?
Ho I guess that works with your definition of G.
But why not just use g_t^2 and remove the diag on the line below?

File: torch/optim/adagrad.py [hunk 3]
```diff
@@ -4,9 +4,30 @@
...
+            &\hspace{5mm}\theta_t \leftarrow
+                \theta_{t-1}-\gamma \mathrm{diag}(G)^{-\frac{1}{2}}g_t                           \\
```
Comment (albanD) [inline torch/optim/adagrad.py:24]: nit missing epsilon

File: torch/optim/adagrad.py (hunk 1 above)
Comment (iramazanli) [reply torch/optim/adagrad.py]: update is done

File: torch/optim/adagrad.py (hunk 2 above)
Comment (iramazanli) [reply torch/optim/adagrad.py]: updated to the shape your recommended

**Review by albanD — APPROVED** [review]
